Refactor flake8 handling and tool compatibility

trws opened the pull request:
This PR does three related things to try to improve developer tooling quality of life:

1. Adds new options to `.flake8` so it applies the rules of both `.flake8` and `.flake_package` based on paths in the repository.
2. Adds a re-factoring of the `spack flake8` logic into a flake8 plugin so using flake8 directly, or through editor or language server integration, only reports errors that `spack flake8` would.
3. Allows star import of `spack.pkgkit` in packages, since this is now the thing that needs to be imported for completion to work correctly in package files, it's nice to be able to do that.

I'm sorely tempted to sed over the whole repository and put `from spack.pkgkit import *` in every package, but at least being allowed to do it on a per-package basis helps.

As an example of what the result of this is:

```
~/Workspace/Projects/spack/spack develop* ⇣
❯ flake8 --format=pylint ./var/spack/repos/builtin/packages/kripke/package.py
./var/spack/repos/builtin/packages/kripke/package.py:6: [F403] 'from spack.pkgkit import *' used; unable to detect undefined names
./var/spack/repos/builtin/packages/kripke/package.py:25: [E501] line too long (88 > 79 characters)

~/Workspace/Projects/spack/spack refactor-flake8*
1 ❯ flake8 --format=spack ./var/spack/repos/builtin/packages/kripke/package.py

~/Workspace/Projects/spack/spack refactor-flake8*
❯ flake8 ./var/spack/repos/builtin/packages/kripke/package.py
```

trws committed:
qa/flake8: update .flake8, spack formatter plugin

Adds:
* Modern flake8 settings for per-path/glob error ignores, allows
  packages to use the same `.flake8` as the rest of spack
* A spack formatter plugin to flake8 that implements the behavior of
  `spack flake8` for direct invocations.  Makes integration with
  developer tooling nicer, linting with flake8 reports only errors that
  `spack flake8` would report.  Using pyls and pyls-flake8, or any other
  non-format-dependent flake8 integration, now works with spack's rules.

trws committed:
qa/flake8: allow star import of spack.pkgkit

To get working completion of directives and spack components it's
necessary to import the contents of spack.pkgkit.  At the moment doing
this makes flake8 displeased.  For now, allow spack.pkgkit and spack
both, next step is to ban spack * and require spack.pkgkit *.

adamjstewart commented:
My only concern is that now there is a lot of duplication between `lib/spack/spack/cmd/flake8.py` and `share/spack/qa/flake8_formatter.py`. Is it possible to simplify `spack flake8` to use the formatter you added? If we do, then `spack flake8` will be identical to `flake8` except that it detects which files have changed. That would allow us to test it as well. If we're adding Python files to `share/spack/qa`, I think we might need to explicitly enable coverage on them.

trws commented:
Absolutely.  I actually started to do it but decided to put this up this way first in case we wanted to preserve that for some reason.  The only reason I can think of we would want to do that is that the plugin system is *relatively* newish to flake8, but since CI runs a new version, and it's only a pip install away I'm not sure that matters so much.  I can refactor the command on top of this shortly.

As to the placement, I put it there just to keep it with the qa stuff, but we can pretty easily move it elsewhere.

adamjstewart commented:
How new is _newish_? Regardless, I'm still in favor of it for the same reasons you outlined. The only potential problem/annoyance is if we introduce a change that means `spack flake8` doesn't work with Python 2.6 anymore.

adamjstewart commented:
> CI runs a new version

`spack flake8` CI runs a new version, but we do have unit tests in `lib/spack/spack/test/cmd/flake8.py` that still need to work.

trws commented:
Support for this kind of local, uninstalled, plugin was added in 3.5.0 which was apparently released in October of 2017.

How do the spack flake8 tests currently work, since flake8 doesn't support python 2.6?

trws commented:
Rebased.

adamjstewart commented:
> How do the spack flake8 tests currently work, since flake8 doesn't support python 2.6?

It looks like we use a `skipif` for Python 2.6, so as long as the latest version of flake8 support Python 2.7 then we're good.

trws commented:
Cool, it claims to so I'll put together that refactor and see what the CI thinks of it.

trws committed:
first cut at refactoring spack flake8

This version still copies all of the files to be checked as befire, and
some other things that probably aren't necessary, but it relies on the
spack formatter plugin to implement the ignore logic.

trws commented:
That was a little nastier than I thought because of `F811`.  I think I have a reliable workaround for the issue in there, but it's a little bit gnarly.

trws committed:
keep flake8 from rejecting itself

trws committed:
remove separate packages flake8 config

trws committed:
fix failures from too many files

I ran into this in the PR converting pkgkit to std.  The solution in
that branch does not work in all cases as it turns out, and all the
workarounds I tried to use generated configs to get a single invocation
of flake8 with a filename optoion to work failed.  It's an astonishingly
frustrating config option.

Regardless, this removes all temporary file creation from the command
and relies on the plugin instead.  To work around the huge number of
files in spack and still allow the command to control what gets checked,
it scans files in batches of 100.  This is a completely arbitrary number
but was chosen to be safely under common line-length limits.  One
side-effect of this is that every 100 files the command will produce
output, rather than only at the end, which doesn't seem like a terrible
thing.

trws committed:
appease python2

Apparently zip_longest didn't exist in python 2.7, nor did commas at the
end of argument lists.

trws committed:
format import in flake8

trws committed:
appease vermine

trws committed:
update completion to remove flake8 --keep-temp

trws committed:
add an extra space, because apparently comments need two

tgamblin commented on .flake8:
```diff
@@ -30,3 +30,36 @@
+# - F821: undefined name `name`
+#
+per-file-ignores =
+  var/spack/repos/*/package.py:F405,F821
```
this is awesome
